larq_compute_engine/mlir/python/common.cc

```diff
@@ -0,0 +1,141 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+Modifications copyright (C) 2021 Larq Contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "common.h"
+
+#include <exception>
+
+#include "larq_compute_engine/mlir/tf_tfl_passes.h"
+#include "larq_compute_engine/mlir/tf_to_tfl_flatbuffer.h"
+#include "larq_compute_engine/mlir/transforms/passes.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/Pass/Pass.h"
+#include "pybind11/pybind11.h"
+#include "tensorflow/compiler/mlir/lite/transforms/passes.h"
+#include "tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/platform/status.h"
+
+namespace tensorflow {
+
+LCETarget GetLCETarget(const std::string& target_str) {
+  if (target_str == "arm") {
+    return LCETarget::ARM;
+  } else if (target_str == "xcore") {
+    return LCETarget::XCORE;
+  } else {
+    throw std::runtime_error("Invalid target.");
+  }
+}
+
+Status GetNumInputs(mlir::OwningModuleRef* module, int* num_inputs) {
+  *num_inputs = 0;
+  mlir::FuncOp entry_function = nullptr;
+  for (auto func : module->get().getOps<mlir::FuncOp>()) {
+    if (auto tf_attrs =
+            func->getAttrOfType<mlir::DictionaryAttr>("tf.entry_function")) {
+      // TODO(jaesung): There could be multiple entry functions. Let's handle
+      // such cases if there are any needs for that.
+      if (entry_function != nullptr) {
+        return errors::InvalidArgument(
+            "There should be only one tf.entry_function");
+      }
+      entry_function = func;
+    }
+  }
+  if (entry_function == nullptr) {
+    return errors::InvalidArgument("no tf.entry_function found");
+  }
+
+  // Get the list of input Op names from the function attribute.
+  mlir::DictionaryAttr tf_attrs =
+      entry_function->getAttrOfType<mlir::DictionaryAttr>("tf.entry_function");
+  llvm::SmallVector<llvm::StringRef, 4> function_input_names;
+  auto input_attr = tf_attrs.get("inputs");
+  if (!input_attr) {
+    return errors::InvalidArgument("no inputs attribute found");
+  }
+  auto input_names = input_attr.cast<mlir::StringAttr>().getValue();
+  input_names.split(function_input_names, ",");
+  *num_inputs = function_input_names.size();
+  return Status::OK();
+}
+
+pybind11::bytes ConvertMLIRModuleToTFLiteFlatBuffer(
```

Comment:
This function body was copied from `saved_model_tfl_flatbuffer.cc` and then adjusted in the following ways:
1) The `should_quantize` option was added to make it work with `graphdef_tfl_flatbuffer.cc` as well.
2) The `mark_as_post_training_quant` option was added to make it work with `graphdef_tfl_flatbuffer.cc` as well.